src/libstd/alloc.rs

```diff
@@ -133,60 +136,110 @@ pub use alloc_crate::alloc::*;
-        } else {
-            unsafe {
-                NonNull::new(GlobalAlloc::alloc_zeroed(self, layout))
-                    .ok_or(AllocErr)
-                    .map(|p| (p, layout.size()))
-            }
+    unsafe fn dealloc(&mut self, ptr: NonNull<u8>, layout: Layout) {
+        if layout.size() != 0 {
+            GlobalAlloc::dealloc(self, ptr.as_ptr(), layout)
         }
     }
 
     #[inline]
-    unsafe fn dealloc(&mut self, ptr: NonNull<u8>, layout: Layout) {
-        if layout.size() != 0 {
-            GlobalAlloc::dealloc(self, ptr.as_ptr(), layout)
+    unsafe fn grow(
+        &mut self,
+        ptr: NonNull<u8>,
+        layout: Layout,
+        new_size: usize,
+        placement: ReallocPlacement,
+        init: AllocInit,
+    ) -> Result<(NonNull<u8>, usize), AllocErr> {
+        let old_size = layout.size();
+        debug_assert!(
+            new_size >= old_size,
+            "`new_size` must be greater than or equal to `layout.size()`"
+        );
+
+        // `new_size == 0` implies `old_size == 0`
+        if old_size == new_size || new_size == 0 {
+            return Ok((ptr, new_size));
+        }
+
+        match placement {
+            Unspecified => {
+                // `new_size` must be greater than or equal to `old_size` due to safety constraint
+                // `new_size` == `old_size` was catched before
+                intrinsics::assume(new_size > old_size);
+                if old_size == 0 {
+                    self.alloc(Layout::from_size_align_unchecked(new_size, layout.align()), init)
+                } else {
+                    let ptr =
+                        NonNull::new(GlobalAlloc::realloc(self, ptr.as_ptr(), layout, new_size))
+                            .ok_or(AllocErr)?;
+
+                    // `GlobalAlloc` doesn't have `realloc_zeroed`, emulate it
+                    if init == Zeroed {
+                        ptr.as_ptr().add(old_size).write_bytes(0, new_size - old_size)
+                    }
+
+                    Ok((ptr, new_size))
+                }
+            }
+            InPlace => Err(AllocErr),
         }
     }
 
     #[inline]
-    unsafe fn realloc(
+    unsafe fn shrink(
         &mut self,
         ptr: NonNull<u8>,
         layout: Layout,
         new_size: usize,
+        placement: ReallocPlacement,
     ) -> Result<(NonNull<u8>, usize), AllocErr> {
-        match (layout.size(), new_size) {
-            (0, 0) => Ok((layout.dangling(), 0)),
-            (0, _) => self.alloc(Layout::from_size_align_unchecked(new_size, layout.align())),
-            (_, 0) => {
-                self.dealloc(ptr, layout);
-                Ok((layout.dangling(), 0))
+        let old_size = layout.size();
+        debug_assert!(
+            new_size <= old_size,
+            "`new_size` must be smaller than or equal to `layout.size()`"
+        );
+
+        // `old_size == 0` implies `new_size == 0`
+        if old_size == new_size || old_size == 0 {
```

Comment:
```suggestion
        if old_size == new_size {
```
See `impl AllocRef for Global`. This is redundant.